### PR TITLE
SONAR-9875 Add branch= param to /api/navigation/component

### PR DESCRIPTION
**Pull request by janos-ss**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/ui/ws/ComponentAction.java [hunk 1]
```diff
@@ -281,7 +290,7 @@ private void writeBreadCrumbs(JsonWriter json, DbSession session, ComponentDto c
 
     for (ComponentDto c : breadcrumb) {
       json.beginObject()
-        .prop("key", c.getDbKey())
+        .prop("key", c.getKey())
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/ui/ws/ComponentAction.java]: We should ask @stas-vilchik if the branch field should now also be returned

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java [hunk 2]
```diff
@@ -219,6 +219,50 @@ public void return_component_info_when_snapshot() throws Exception {
...
 
   @Test
+  public void return_component_info_when_file_on_master() throws Exception {
+    init();
+    OrganizationDto organization = dbTester.organizations().insertForKey("my-org2");
+    ComponentDto main = componentDbTester.insertMainBranch(organization, p -> p.setName("Sample"), p -> p.setDbKey("sample"));
+    userSession.addProjectPermission(UserRole.USER, main);
+
+    ComponentDto dirDto = newDirectory(main, "src");
+    componentDbTester.insertComponent(dirDto);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: You can inline dirDto

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java [hunk 3]
```diff
@@ -219,6 +219,50 @@ public void return_component_info_when_snapshot() throws Exception {
...
+
+    ComponentDto fileDto = newFileDto(main, dirDto)
+      .setUuid("abcd")
+      .setName("Main.xoo")
+      .setDbKey("sample:src/Main.xoo");
+    componentDbTester.insertComponent(fileDto);
```
Comment (julienlancelot) [inline server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: You can inline fileDto

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java (hunk 2 above)
Comment (janos-ss) [reply server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: I certainly can. But will it be a good idea? That would mean replacing 3 lines with 1 longer line:

    ComponentDto fileDto = newFileDto(branch, componentDbTester.insertComponent(newDirectory(branch, "src")))

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java (hunk 3 above)
Comment (janos-ss) [reply server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: I use `fileDto.getDbKey()` a few lines below this. Sure I could cram all the `componentDbTester.insertComponent(...).getDbKey()` in there, but I think readability would suffer.

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java (hunk 2 above)
Comment (julienlancelot) [reply server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: Sorry, it was not clear, I was more thinking of : 
`ComponentDto dirDto = componentDbTester.insertComponent(newDirectory(main, "src"))`

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java (hunk 3 above)
Comment (julienlancelot) [reply server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: Same as before, I would write : 
'ComponentDto fileDto = componentDbTester.insertComponent(newFileDto(main, dirDto).set...)'

File: server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java (hunk 2 above)
Comment (janos-ss) [reply server/sonar-server/src/test/java/org/sonar/server/ui/ws/ComponentActionTest.java]: Oh I see, I agree with that. I think "inline" means "replace with implementation", and I interpreted that as eliminating the local variables completely.

File: server/sonar-server/src/main/java/org/sonar/server/ui/ws/ComponentAction.java (hunk 1 above)
Comment (janos-ss) [reply server/sonar-server/src/main/java/org/sonar/server/ui/ws/ComponentAction.java]: @julienlancelot Stas confirmed he doesn't need it.

**Comment by julienlancelot** [comment]
Unfortunately clicking on the permalink on a issue of a branch is still failing with a 404.
The reason is that the link is incorrect : it contains the component dbKey in **project** and **subProject** fields. See for instance : 
```
"issues": [
    {
        "key": "AV7DfbyWRBW-wJqITO1l",
        "rule": "squid:S3776",
        "severity": "CRITICAL",
        "component": "com.sonarsource.branch:sonar-branch-plugin:src/main/java/com/sonarsource/branch/BranchConfigurationLoaderImpl.java",
        "project": "com.sonarsource.branch:sonar-branch:BRANCH:branch-test",
        "subProject": "com.sonarsource.branch:sonar-branch-plugin:BRANCH:branch-test"
}
]
```
You need to fix org.sonar.server.issue.ws.SearchResponseFormat#formatIssue:167

**Pull request merged**